src/rust/src/x509/verify.rs

```diff
@@ -0,0 +1,226 @@
+    type Key = CryptographyResult<pyo3::Py<pyo3::PyAny>>;
+
+    fn public_key(&self, cert: &Certificate<'_>) -> Self::Key {
+        pyo3::Python::with_gil(|py| -> Self::Key {
+            // This makes an unnecessary copy. It'd be nice to get rid of it.
+            let spki_der = pyo3::types::PyBytes::new(py, &asn1::write_single(&cert.tbs_cert.spki)?);
+            Ok(py
+                .import(pyo3::intern!(
+                    py,
+                    "cryptography.hazmat.primitives.serialization"
+                ))?
+                .getattr(pyo3::intern!(py, "load_der_public_key"))?
+                .call1((spki_der,))?
+                .into())
+        })
+    }
+
+    fn is_signed_by(&self, cert: &Certificate<'_>, key: Self::Key) -> bool {
+        let key = match key {
+            Ok(key) => key,
+            Err(_) => return false,
+        };
+
+        pyo3::Python::with_gil(|py| -> CryptographyResult<()> {
+            sign::verify_signature_with_signature_algorithm(
+                py,
+                key.as_ref(py),
+                &cert.signature_alg,
+                cert.signature.as_bytes(),
+                &asn1::write_single(&cert.tbs_cert)?,
+            )
+        })
+        .is_ok()
+    }
+}
+
+#[pyo3::pyclass(name = "Profile", module = "cryptography.hazmat.bindings._rust.x509")]
+enum PyProfile {
+    RFC5280,
+    WebPKI,
+}
+
+// NOTE: pyO3 classes can't be generic over a trait, which our profile API requires.
+// Our workaround is to unfold each profile variant in this enum, which produces
+// a small amount of (internal) duplication in exchange for the monomorphism that
+// pyO3 needs.
+enum FixedPolicy {
+    RFC5280(Policy<PyCryptoOps, RFC5280>),
+    WebPKI(Policy<PyCryptoOps, WebPKI>),
+}
+
+impl FixedPolicy {
+    fn name(&self) -> Option<&str> {
+        match self {
+            FixedPolicy::RFC5280(policy) => policy.name.as_ref().map(|n| n.as_str()),
+            FixedPolicy::WebPKI(policy) => policy.name.as_ref().map(|n| n.as_str()),
+        }
+    }
+
+    fn time(&self) -> &asn1::DateTime {
+        match self {
+            FixedPolicy::RFC5280(policy) => &policy.validation_time,
+            FixedPolicy::WebPKI(policy) => &policy.validation_time,
+        }
+    }
+}
+
+#[pyo3::pyclass(name = "Policy", module = "cryptography.hazmat.bindings._rust.x509")]
+struct PyPolicy(FixedPolicy);
+
+#[pyo3::pymethods]
+impl PyPolicy {
+    #[getter]
+    fn name<'p>(&self, py: pyo3::Python<'p>) -> Option<&'p pyo3::types::PyString> {
+        self.0
+            .name()
+            .map(|name| pyo3::types::PyString::new(py, name))
+    }
+
+    #[getter]
+    fn time<'p>(&self, py: pyo3::Python<'p>) -> pyo3::PyResult<&'p pyo3::PyAny> {
+        datetime_to_py(py, self.0.time())
+    }
+}
+
+#[pyo3::prelude::pyfunction]
+fn create_policy(
+    py: pyo3::Python<'_>,
+    profile: &pyo3::PyAny,
+    name: Option<&pyo3::types::PyString>,
+    time: Option<&pyo3::PyAny>,
+) -> pyo3::PyResult<PyPolicy> {
```

Comment:
Hmm, this is an annoyance: our `Policy` type currently has an interior lifetime because it holds onto a borrowing `asn1::IA5String`, but that doesn't mesh nicely with the signature here (we expect to return a fully owned `PyPolicy`, and `pyo3` doesn't play nicely with interior lifetimes).

We could hack around that with a self-referencing `OwnedPolicy` a la `OwnedCertificate`, but that feels pretty heavy for just a single (relatively short) string. It might also not work so well once we extend `Policy` to support SANs other than DNS names, since not all of those types may be borrowing (e.g., for IP addresses we'll probably just reuse `std::net::IpAddr`).

Reply:
I'd be inclined to just suggest using self_cell to do the self-referential thing where the PyString owns the memory and Policy borrows it. It's not much overhead.

One other thing to be thinking about: this is really specific to a client's validation policy (i.e., verifying a server's cert). For the inverse (a server verifying a client cert) there's a different structure: you don't pass a name to verify, instead you get back a list of valid names.